### ISSUE TITLE
http(s).request() not cancelled when destroyed, and leaking memory

### ISSUE THREAD

**Issue by oguimbal** [oven-sh/bun#10665]
### What version of Bun is running?

1.1.6+e58d67b46

### What platform is your computer?

Darwin 23.1.0 arm64 arm

### What steps can reproduce the bug?

1) Write this in a `test.js` file
```javascript
import http from "https";

// just some file that will take forever to download
const bigFile =
  `https://ash-speed.hetzner.com/10GB.bin`;

let req = null;
let downloaded = 0;
try {
  // launch a GET query
  await new Promise((resolve) => {
    req = http.request(bigFile, (res) => {
      resolve(res);
      res
        .on("data", (chunk) => {
          downloaded += chunk.length;
        })
        .resume();
    });
    req.end();
  });

  //just download a tiny bit
  await new Promise((resolve) => setTimeout(resolve, 10));
} finally {
  // destroy the stream (should stop the download)
  console.log("Destroying stream");
  req.destroy();
}

// check if the download is still progressing
let int;
let previousDownloaded = 0;
int = setInterval(() => {
  if (downloaded === previousDownloaded) {
    console.log("No progress => should shutdown");
    clearInterval(int);
  } else {
    console.log("Progres", downloaded);
    previousDownloaded = downloaded;
  }
}, 1000);

```

2) Check that it works in node:
```bash
> node test.js
Destroying stream
Downloaded 63680
No progress => should shutdown
```
(the program exits)

3) Try it in bun:
```bash
> bun test.js
Destroying stream
Downloaded 6355154
Downloaded 36321490
Downloaded 66459858
Downloaded 97974482
Downloaded 133183698
Downloaded 168376530
Downloaded 203159762
......
```

### What is the expected behavior?

the `request.destroy()` should stop the download.

### What do you see instead?

1) The download does not stop in bun.
2) The program is kept alive by this query (it should stop)
3) If the response is piped to a closed stream, it looks like this leaks memory (we're seing huge increases in memory after request cancellations, until the process cashes with OOM)

### Additional information

We're a company, and we use bun in prod 😇

**Comment by oguimbal** [comment]
Additional info:

If you use this variant in bun, you will see that bun leaks memory (seems like the cancelled download never stops, and is buffered somewhere... memory is never released)

```javascript
import { shrink } from "bun";
import http from "https";
import fs from "fs";

// just some file that will take forever to download
const bigFile = "https://ash-speed.hetzner.com/10GB.bin";

// GC often to be sure we're leaking memory
setInterval(() => {
  Bun.gc(true);
  shrink();
}, 1000);

let req = null;
const f = fs.createWriteStream("10GB.bin");

// download file
await new Promise(resolve => {
  req = http.request(bigFile, res => {
    resolve();
    res.pipe(f);
  });
  req.end();
});

//just download a tiny bit
await new Promise((resolve) => setTimeout(resolve, 10));

req.destroy(); // should stop the download

// 👉 close the piped file => this triggers a memory leak !
f.close();
```

**Comment by Jarred-Sumner** [comment]
The fix for this will be included in Bun v1.1.7, which releases today. Thanks for your patience.

if you want to try it now, you can download the canary release of Bun:
```sh
bun upgrade --canary
```
